Signed-off-by: Sven Rebhan <odinshorse@googlemail.com>

Always try to prepend the sysroot prefix to absolute filenames first.

http://bugs.gentoo.org/275666
http://sourceware.org/bugzilla/show_bug.cgi?id=10340

 (modified by rrw@kynesim.co.uk 2016-12-30 to apply to binutils 2.27)


--- binutils-2.27-orig/ld/ldfile.c	2016-08-03 08:36:54.000000000 +0100
+++ binutils-2.27/ld/ldfile.c	2016-12-30 14:24:24.846824899 +0000
@@ -336,13 +336,19 @@
      directory first.  */
   if (!entry->flags.maybe_archive)
     {
-      if (entry->flags.sysrooted && IS_ABSOLUTE_PATH (entry->filename))
+      /* For absolute pathnames, try to always open the file in the
+	 sysroot first. If this fails, try to open the file at the
+	 given location.  */
+      entry->flags.sysrooted = is_sysrooted_pathname (entry->filename);
+      if (!entry->flags.sysrooted && IS_ABSOLUTE_PATH (entry->filename) && 
+          ld_sysroot)
 	{
 	  char *name = concat (ld_sysroot, entry->filename,
 			       (const char *) NULL);
 	  if (ldfile_try_open_bfd (name, entry))
 	    {
 	      entry->filename = name;
+              entry->flags.sysrooted = TRUE;
 	      return TRUE;
 	    }
 	  free (name);
